### ISSUE TITLE
Google Play Music update breaks Nuvola [$15]

### ISSUE THREAD

**Issue by daniel-centore** [tiliado/nuvolaplayer#100]
The Fix
-------------

The latest build of Nuvola Player (both beta and devel) and Google Play Music web app script 5.3 contain a fix for the issue. Unlike the hotfix announced earlier, the final fix is enabled by default.

Google Play Music seems to play music fine now. Remaining issues will be addressed in separate tickets (e.g. #106 Google Play Music shows an incomplete queue). Thanks for your reports, patience and donations.

---

Jiri Janousek, Nuvola Player Developer • My work on Nuvola Player and user support are [funded by Nuvola Patrons](https://tiliado.eu/nuvolaplayer/funding/). [Make a donation](https://tiliado.eu/nuvolaplayer/funding/) to keep the project going. Thank you!

Tasks
-------

You can [support my work](https://tiliado.eu/nuvolaplayer/funding/)  on this issue.

- [x] Implement final fix based on `ResourceRequest` signal.
- [ ] Add and use API `Nuvola.Browser.reloadWithoutCache()`
- [x] Investigate `listen.js:413:170 JS ERROR: TypeError: null is not an object (evaluating 'a[IG]')`.

The Issue
-------------

Google Play Music updated their layout today, and as a result of this Nuvola Player (both the 2.x branch and 3.0 Beta, which I am currently running) no longer works correctly. It is impossible to get any songs to play, to open the sidebar, etc. Both versions were working fine this morning before the layout update.

![image](https://cloud.githubusercontent.com/assets/569864/7641133/a5949b4e-fa55-11e4-86e9-ae14f5971c12.png)

**Comment by fenryxo** [comment]
Hello. Thanks for your report. It seems to be a rather serious issue this time as there are JavaScript errors produced by the Google Play code :-( Will try to reach WebKitGTK developers and Google Play Music support.

![google_play_music_error](https://cloud.githubusercontent.com/assets/853706/7647978/6f297880-fadd-11e4-9acb-1101b7afadbf.png)

```
[Error] CustomError: Error in protected function:
null is not an object (evaluating 'this.ioa.aNa')
	c (listen.js, line 3101)
	xna (listen.js, line 917)
	(anonymous function) ([native code], line 0)
[Warning] Unexpected CSS token: : (listen, line 322, x8)
[Error] Refused to get unsafe header "Location"
	(anonymous function) (webcomponents.js, line 5848)
[Error] TypeError: A WeakMap cannot have a non-object key
	get (webcomponents.js, line 167)
	getWrapperConstructor (webcomponents.js, line 167)
	getWrapperConstructor (webcomponents.js, line 169)
	wrap (webcomponents.js, line 293)
	firstChild (webcomponents.js, line 1931)
	firstElementChild (webcomponents.js, line 2249)
	children (webcomponents.js, line 2264)
	addedNodes (webcomponents.js, line 5977)
	handler (webcomponents.js, line 5968)
	(anonymous function) ([native code], line 0)
	notifyObservers (webcomponents.js, line 587)
	handle (webcomponents.js, line 532)
[Error] CustomError: Error in protected function:
A WeakMap cannot have a non-object key
	(anonymous function) (listen.js, line 3101)
```

**Comment by fenryxo** [comment]
Reported to Google, please up-vote: [Bug: Google Play Music recently broken in WebKitGTK+ based web browsers](https://productforums.google.com/forum/#!topic/play/U98y_b4Uxk0;context-place=forum/play)

**Comment by fenryxo** [comment]
Same issue in Radiant Player for Mac: kbhomes/radiant-player-mac#318

**Comment by jxn** [comment]
I'm not sure if this is relevant, but here's something related in google-music-webkit with a resolution (?): twolfson/google-music-webkit#6

**Comment by fenryxo** [comment]
There seems to be a bug in Polymer's webcomponents.js polyfill - https://github.com/kbhomes/radiant-player-mac/issues/318#issuecomment-104070925:

> [...] the way I fixed the Flash issue was to dig into the Polymer source, where I found that there was a discrepancy/bug that appears in Cocoa WebViews but not in Safari. More information here: kbhomes/radiant-player-mac@7cbdb20

Will try to get this fix to Nuvola Player.

**Comment by fenryxo** [comment]
> I'm not sure if this is relevant, but here's something related in google-music-webkit with a resolution (?): twolfson/google-music-webkit#6

That issue is about how to make Google Player Music play with HTML5 Audio in Chromium-based browsers. It doesn't apply to Nuvola Player, because it is WebKitGTK-based.

**Comment by DidierLoiseau** [comment]
Just wondering: shouldn't we also comment on [the Google Play Forum thread](https://productforums.google.com/forum/#!topic/play/U98y_b4Uxk0;context-place=forum/play) mentionned by @fenryxo? It has 100+ votes and does not seem to get any attention.

Also, I could not find a way to quickly view a thread's votes or to sort threads by votes, so I wonder if the vote system _really_ deserves a purpose outside the poster's satisfaction and “reputation” (as I understand @fenryxo reached _Level 3_ with this single post?).

**Comment by fenryxo** [comment]
The fixed version of `webcomponents.js` from kbhomes/radiant-player-mac@7cbdb20 does work! We have a hotfix for the flash bug (06e8803), but the integration script needs an update. Will continue to work on this issue in early June as funds for May have been already spent.

**Comment by fenryxo** [comment]
I've updated the header of this issue with instructions how to enable a hotfix available in the latest builds of Nuvola Player. I would like to thank to Ivan Naitana who sponsored development of this hotfix.The final fix is in development.

---

Jiri Janousek, Nuvola Player Developer • My work on Nuvola Player and user support are [funded by Nuvola Patrons](https://tiliado.eu/nuvolaplayer/funding/). [Make a donation](https://tiliado.eu/nuvolaplayer/funding/) to keep the project going. Thank you!

**Comment by KruyKaze** [comment]
Switched to nuvola3 ( on Arch ) and I can confirm the fix works!

But I have lost the command line which is essencial to me. This is what I get:
"nuvolaplayer-client: command not found"
Do I have to install that seperately?
Thanks

**Comment by fenryxo** [comment]
Hi @KruyKaze . I've moderated your comment to keep conversation on topic. Please create a new issue or [a thread in users forum](https://groups.google.com/d/forum/nuvola-player-users) if you have additional questions.

**Comment by fastcat** [comment]
I had to reload the interface a few times, but eventually this hotfix worked.

**Comment by orschiro** [comment]
For me it does almost work. There is still a white box shown in the bottom-left corner.

Multiple reloads of the interface do not remove it.

![image](https://cloud.githubusercontent.com/assets/170947/7792227/440fe5a6-02af-11e5-8954-d262e5bdc344.png)

**Comment by sasha-** [comment]
It mostly works. As indicated in the first post ou have to run player like this:
NUVOLA_ISSUE_100_HOTFIX=true nuvolaplayer3 -v -a google_play_music &

I find that I can select radio station and start the music but cannot see queue. Clicking "Open queue" brings up just the bottom of the queue box, an arrow pointing down, but there is no track list. Still, It is better than having to run Chrome in the background.

**Comment by fenryxo** [comment]
Hi @fastcat. Maybe the broken Google's script is cached and the page needs to be reloaded multiple times before the cached script is replaced with the fixed one.

Hi @orschiro. Could you post terminal output of command `NUVOLA_ISSUE_100_HOTFIX=true nuvolaplayer3 -v -a google_play_music` after the website is loaded?

**Comment by fastcat** [comment]
That's possible ... also possible is that I hit "Home" once instead of reload, and that triggered a different behavior.  Perhaps the cached URL was no longer valid and not redirecting properly to valid one.  At any rate it's working for me now :)

**Comment by orschiro** [comment]
@fenryxo 

It still leaves me with a half-loaded screen:

![image](https://cloud.githubusercontent.com/assets/170947/7807273/5cb24cc6-038a-11e5-9ece-c6b7d93793b7.png)

```
~ ❯❯❯ NUVOLA_ISSUE_100_HOTFIX=true nuvolaplayer3 -v -a google_play_music
[Master:INFO       Gtk] Failed to load module "overlay-scrollbar"
[Master:WARNING  <unknown>] Couldn't connect to accessibility bus: Failed to connect to socket /tmp/dbus-AjqooWER7F: Connection refused
[Master:INFO     Nuvola] WebAppRegistry.vala:89: Using web app Google Play Music, version 5.2, data dir /usr/share/nuvolaplayer3/web_apps/google_play_music
[Runner:INFO       Gtk] Failed to load module "overlay-scrollbar"
[Runner:WARNING  <unknown>] Couldn't connect to accessibility bus: Failed to connect to socket /tmp/dbus-AjqooWER7F: Connection refused
Gtk-Message: Failed to load module "overlay-scrollbar"

** (WebKitWebProcess:16334): WARNING **: Couldn't connect to accessibility bus: Failed to connect to socket /tmp/dbus-AjqooWER7F: Connection refused
[Worker:DEBUG    Nuvola] main.vala:201: Page 1 created for (null)
[Worker:INFO     Nuvola] main.vala:207: Nuvola Issue #100 hotfix enabled. https://github.com/tiliado/nuvolaplayer/issues/100
[Worker:DEBUG    Nuvola] main.vala:125: Window object cleared: https://play.google.com/music/listen#/pl/AMaBXyncdZe9KmPxuacWksOeHMS44B_oV1oZLL6fN32rs2JZ12X4RKKujpnkNqk7vK0A8Nxbaw3PqdLmj7Y6wSRiC3CpTdfZYA%3D%3D, 0x10d94a0, 0x10d9130, 0x7f5e3c13f9b0
[Worker:INFO     Nuvola] main.vala:222: Broken web components: https://ssl.gstatic.com/play/music/fe/b787edead24af83f0b329116750567eb/webcomponents.js
[Worker:INFO     Nuvola] main.vala:224: Fixed web components: https://raw.githubusercontent.com/kbhomes/radiant-player-mac/master/radiant-player-mac/js/webcomponents.js
[Runner:INFO     Nuvola] AppRunnerController.vala:236: Logged in as  ()
https://play.google.com/music/listen#/pl/AMaBXyncdZe9KmPxuacWksOeHMS44B_oV1oZLL6fN32rs2JZ12X4RKKujpnkNqk7vK0A8Nxbaw3PqdLmj7Y6wSRiC3CpTdfZYA%3D%3DJS ERROR: Refused to get unsafe header "Location"
https://ssl.gstatic.com/play/music/fe/b787edead24af83f0b329116750567eb/listen.js:413:170JS ERROR: TypeError: null is not an object (evaluating 'a[IG]')
Gtk-Message: Failed to load module "overlay-scrollbar"
Gtk-Message: Failed to load module "canberra-gtk-module"
[Runner:INFO     Nuvola] config.vala:260: Config saved to /home/orschiro/.config/nuvolaplayer3/apps_data/google_play_music/config.json
```

**Comment by fenryxo** [comment]
@orschiro It sometimes happens. It usually helps to click Home button and then Reload button. The cause is probably `listen.js:413:170 JS ERROR: TypeError: null is not an object (evaluating 'a[IG]')`. This error needs more investigation (scheduled for June 2015).

**Comment by orschiro** [comment]
> It sometimes happens. It usually helps to click Home button and then Reload button

Yes it does, thanks very appreciated!

**Comment by charlwillia6** [comment]
Yes!  Thanks for the quick turnaround on this fix!  Awesome.

**Comment by fenryxo** [comment]
The latest build of Nuvola Player (3.0.0~201506062028, both beta and devel) and Google Play Music web app script 5.3 contain a final fix for the issue. Unlike the hotfix announced earlier, the final fix is enabled by default. Please report whether you have any issues.

**Comment by mathus13** [comment]
new build does not display the queue

![image](https://cloud.githubusercontent.com/assets/459913/8061544/5b39b6bc-0e99-11e5-9410-35ebf9d77a77.png)

**Comment by fastcat** [comment]
Queue display works for me in the new build -- displays automatically when I start playing a stream, and clicking the icon toggles it on and off.

**Comment by MRamiro** [comment]
I can confirm that the new build shows an incomplete queue. Notice the presence of a white line and triangle where the queue window should be:

![captura de pantalla de 2015-06-09 12 42 14](https://cloud.githubusercontent.com/assets/3259212/8065133/118ff948-0ea5-11e5-8912-c505aa9e81b1.png)

Queue worked fine in the previous build while using the hotfix.

Regards.

**Comment by RCheesley** [comment]
Firstly thanks for the fix, the update is now allowing music to play, but I'm also getting the missing queue as mentioned by @MRamiro in the bottom right corner - I'm running Kubuntu 14.04 with version 3.0.0+201506062028.00f51f7 and revision 00f51f7bac10b2e79bdc

![google-play-queue](https://cloud.githubusercontent.com/assets/2930593/8079988/8b2fe57c-0f60-11e5-83ec-4197a0a4d1dc.png)

**Comment by dvapelnik** [comment]
I have a same bug
![laptop_20150610__15 05 23](https://cloud.githubusercontent.com/assets/1562244/8081978/4373c196-0f82-11e5-96bf-0b9410559366.png)
![laptop_20150610__15 05 36](https://cloud.githubusercontent.com/assets/1562244/8081981/4a216994-0f82-11e5-9a7e-f5f9a409d1ca.png)
![laptop_20150610__15 06 56](https://cloud.githubusercontent.com/assets/1562244/8081985/6222ea04-0f82-11e5-9c4a-2e0467b66771.png)

**Comment by fenryxo** [comment]
@mathus13 @MRamiro @RCheesley @dvapelnik I've created a new ticket for the missing queue issue - tiliado/nuvolaplayer#106. Please post there a complete bug details as described in [Bug Reporting Guidelines](https://github.com/tiliado/nuvolaplayer/wiki/Bug-Reporting-Guidelines), especially the debugging output.

@dvapelnik Please don't post screenhots about Nuvola Player and your distribution versions, but write it as text.

**Comment by fenryxo** [comment]
I'm closing this ticket as Google Play Music seems to play music fine now. Remaining issues will be addressed in separate tickets (e.g. #106 Google Play Music shows an incomplete queue). Thanks for your reports, patience and donations.

---

Jiri Janousek, Nuvola Player Developer • My work on Nuvola Player and user support are [funded by Nuvola Patrons](https://tiliado.eu/nuvolaplayer/funding/). [Make a donation](https://tiliado.eu/nuvolaplayer/funding/) to keep the project going. Thank you!

**Comment by t3mujin** [comment]
Hi all, is it only me or this started happening again? Another breaking change on Google Play?

**Comment by litrik** [comment]
Yup, I have noticed it as well. Yesterday everything was working fine, but now it is broken again.
Maybe you should create a new issue (instead of reopening this one)

**Comment by fenryxo** [comment]
Google has replaced a broken file `webcomponents.js` with a broken file `webcomponents.min.js`. It is necessary to adapt our workaround to also substitute the new file. Will take a look at it in [September 2015](https://github.com/tiliado/nuvolaplayer.project/issues/7) as funds for [August 2015](https://github.com/tiliado/nuvolaplayer.project/issues/5) have been exhausted.

**Comment by fenryxo** [comment]
Thanks for the bounty. I've just fixed the issue in git and I'm building packages now. For those who build Nuvola Player from source (AUR): make sure to update diorite, nuvolaplayer and nuvola-app-google-play-music as well.

**Comment by fenryxo** [comment]
Fixed in Google Play Music script 5.6 :-)
